splice_machine/src/main/java/com/splicemachine/pipeline/foreignkey/ForeignKeyParentInterceptWriteHandler.java

```diff
@@ -80,135 +79,68 @@ private boolean isForeignKeyInterceptNecessary(KVPair.Type type) {
-                readUncommittedFilter = new SimpleTxnFilter(Long.toString(indexConglomerateId), ((ActiveWriteTxn) ctx.getTxn()).getReadUncommittedActiveTxn(), NoOpReadResolver.INSTANCE, SIDriver.driver().getTxnStore());
-
+    public void next(KVPair mutation, WriteContext ctx) {
+        if (isForeignKeyInterceptNecessary(mutation.getType())) {
+            if(failed) {
+                ctx.notRun(mutation);
+                return;
             }
-            else if (ctx.getTxn() instanceof WritableTxn) {
-                readCommittedFilter = new SimpleTxnFilter(Long.toString(indexConglomerateId), ((WritableTxn) ctx.getTxn()).getReadCommittedActiveTxn(), NoOpReadResolver.INSTANCE, SIDriver.driver().getTxnStore());
-                readUncommittedFilter = new SimpleTxnFilter(Long.toString(indexConglomerateId), ((WritableTxn) ctx.getTxn()).getReadUncommittedActiveTxn(), NoOpReadResolver.INSTANCE, SIDriver.driver().getTxnStore());
+            try {
+                ensureBuffers(ctx);
+            } catch (Exception e) {
+                ctx.failed(mutation, WriteResult.failed(e.getMessage()));
+                return;
             }
-            else
-                throw new IOException("invalidTxn");
-        try(DataScanner scanner = table.openScanner(scan)) {
-            List<DataCell> next;
-            while ((next = scanner.next(-1)) != null && !next.isEmpty()) {
-                readCommittedFilter.reset();
-                readUncommittedFilter.reset();
-                if (hasData(next, readCommittedFilter) || hasData(next, readUncommittedFilter))
-                    return true;
+            for(Action action : actions.values()) {
+                action.next(mutation, ctx);
+                if(action.hasFailed()) {
+                    failed = true;
+                    ctx.failed(mutation, action.getWriteResult());
+                    break;
+                }
             }
-            return false;
-        }catch (Exception e) {
-            throw new IOException(e);
+        }
+        if(!failed) {
+            ctx.success(mutation);
+            ctx.sendUpstream(mutation);
         }
     }
 
-    private boolean hasData(List<DataCell> next, SimpleTxnFilter txnFilter) throws IOException {
-        int cellCount = next.size();
-        for(DataCell dc:next){
-            DataFilter.ReturnCode rC = txnFilter.filterCell(dc);
-            switch(rC){
-                case NEXT_ROW:
-                    return false; //the entire row is filtered
-                case SKIP:
-                case NEXT_COL:
-                case SEEK:
-                    cellCount--; //the cell is filtered
-                    break;
-                case INCLUDE:
-                case INCLUDE_AND_NEXT_COL: //the cell is included
-                default:
-                    break;
+    @Override
+    public void flush(WriteContext ctx) throws IOException {
+        if(failed) {
```

Comment:
Shouldn't we flush the rows that were written before the failure?

Reply:
Wouldn't that interfere with transaction atomicity in case of failure? e.g.
Imagine the parent table `p` has FKs defined on three child tables (`c1`, `c2`, and `c3`), the `DELETE` on the parent key comprises three rows `r1`, `r2`, and `r3`:
- for `r1`: accepted by all children => add mutations `r1c1`, `r1c2`, `r1c3` to respective buffers.
- for `r2`: accepted only by `c1` => generate mutation `r2c1` and add it to its buffer.
Now if we flush previous results we will end up with semi-updated child tables (`r1c1` and `r2c1` for `c1`, `r1c2` for `c2`, and `r1c3` for `c3`).
I am not sure about the internals of the write buffer so maybe in this situation the statement is rolled back properly at some other level.

Reply:
I was thinking of "permissive" writes, where we would want to write some rows and fail others. For a non-permissive write we would rollback the transaction and those written records shouldn't matter. But this is fine for now until we handle permissive writes for FK.